NPE when combining hibernate search and contained resources

New lucene indexing NPEs with contained resources.

**To Reproduce**
In ModelConfig enable:
- enable `indexOnContainedResources`
- and `indexOnContainedResourcesRecursively`
In DaoConfig enable `advancedLuceneIndexing`
POST /Observation, contained Patient
{
  "resourceType": "Observation",
  "contained": [
    {
      "resourceType": "Patient",
      "id": "contained-patient",
      "name": [{
        "family": "Smith"
      }]
    }
  ],
  "subject": { "reference": "#contained-patient" }
}

**Expected behavior**
200 success

** Actual behaviour**

```
Caused by: org.springframework.transaction.TransactionSystemException: Could not commit JPA transaction; nested exception is javax.persistence.RollbackException: Error while committing the transaction
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.doExecuteCallback(HapiTransactionService.java:149) [4 skipped]
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.execute(HapiTransactionService.java:83)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.execute(HapiTransactionService.java:75)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.create(BaseHapiFhirResourceDao.java:212)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.create(BaseHapiFhirResourceDao.java:192)
... snip ...
**Caused by: org.hibernate.search.util.common.SearchException: HSEARCH700083: Exception while building document for entity 'ResourceTable#1335': HSEARCH600039: Invalid field reference for this document element: this document element has path 'sp.subject.general-practitioner', but the referenced field has a parent with path 'sp'.**
```

Hapi Fair 5.7.0-SNAPSHOT-PRE.

--- a/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3371-fix-error-indexing-contained-lucene.yaml
+++ b/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3371-fix-error-indexing-contained-lucene.yaml
@@ -1,0 +1,5 @@
+---
+type: fix
+issue: 3371
+title: "Enhanced Lucene Indexing failed when indexing contained resources.
+   Contained resources are not indexed in Lucene/Elasticsearch, but this no longer causes an exception."

--- a/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneSearchBuilder.java
+++ b/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedLuceneSearchBuilder.java
@@ -59,7 +59,7 @@ public class ExtendedLuceneSearchBuilder {
 				// each and clause may have a different modifier, so split down to the ORs
 				.flatMap(andList -> andList.getValue().stream())
 				.flatMap(Collection::stream)
-				.anyMatch(this::isParamSupported);
+				.anyMatch(this::isParamTypeSupported);
 	}
 
 	/**
@@ -67,7 +67,7 @@ public class ExtendedLuceneSearchBuilder {
 	 *
 	 * NOTE - keep this in sync with addAndConsumeAdvancedQueryClauses() below.
 	 */
-	private boolean isParamSupported(IQueryParameterType param) {
+	private boolean isParamTypeSupported(IQueryParameterType param) {
 		String modifier = StringUtils.defaultString(param.getQueryParameterQualifier(), EMPTY_MODIFIER);
 		if (param instanceof TokenParam) {
 			switch (modifier) {

--- a/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java
+++ b/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java
@@ -15,6 +15,7 @@ import ca.uhn.fhir.jpa.dao.data.IResourceTableDao;
 import ca.uhn.fhir.jpa.entity.TermCodeSystemVersion;
 import ca.uhn.fhir.jpa.entity.TermConcept;
 import ca.uhn.fhir.jpa.entity.TermConceptParentChildLink;
+import ca.uhn.fhir.jpa.model.entity.ModelConfig;
 import ca.uhn.fhir.jpa.model.entity.ResourceTable;
 import ca.uhn.fhir.jpa.search.reindex.IResourceReindexingSvc;
 import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
@@ -51,6 +52,7 @@ import org.hl7.fhir.r4.model.StringType;
 import org.hl7.fhir.r4.model.ValueSet;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.springframework.beans.factory.annotation.Autowired;
@@ -119,15 +121,10 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 	private IBulkDataExportSvc myBulkDataExportSvc;
 	@Autowired
 	private ITermCodeSystemStorageSvc myTermCodeSystemStorageSvc;
-	private boolean myContainsSettings;
 
 	@BeforeEach
 	public void beforePurgeDatabase() {
 		purgeDatabase(myDaoConfig, mySystemDao, myResourceReindexingSvc, mySearchCoordinatorSvc, mySearchParamRegistry, myBulkDataExportSvc);
-		myDaoConfig.setAdvancedLuceneIndexing(true);
-		// we don't support chained or contained yet, but turn it on to test we don't blow up.
-		myDaoConfig.getModelConfig().setIndexOnContainedResources(true);
-		myDaoConfig.getModelConfig().setIndexOnContainedResourcesRecursively(true);
 	}
 
 	@Override
@@ -141,14 +138,16 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 	}
 
 	@BeforeEach
-	public void enableContains() {
-		myContainsSettings = myDaoConfig.isAllowContainsSearches();
+	public void enableContainsAndLucene() {
 		myDaoConfig.setAllowContainsSearches(true);
+		myDaoConfig.setAdvancedLuceneIndexing(true);
 	}
 
 	@AfterEach
 	public void restoreContains() {
-		myDaoConfig.setAllowContainsSearches(myContainsSettings);
+		DaoConfig defaultConfig = new DaoConfig();
+		myDaoConfig.setAllowContainsSearches(defaultConfig.isAllowContainsSearches());
+		myDaoConfig.setAdvancedLuceneIndexing(defaultConfig.isAdvancedLuceneIndexing());
 	}
 
 	@Test
@@ -503,29 +502,46 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 		assertThat(message, toUnqualifiedVersionlessIdValues(myObservationDao.search(map)), containsInAnyOrder(toValues(iIdTypes)));
 	}
 
-	/**
-	 * We were throwing when indexing contained.
-	 * https://github.com/hapifhir/hapi-fhir/issues/3371
-	 */
-	@Test
-	public void ignoreContainedResources_noError() {
-	    // given
-		String json =
-			"{" +
-			"\"resourceType\": \"Observation\"," +
-			"\"contained\": [{" +
-			"\"resourceType\": \"Patient\"," +
-			"\"id\": \"contained-patient\"," +
-			"\"name\": [{ \"family\": \"Smith\"}]" +
-			"}]," +
-			"\"subject\": { \"reference\": \"#contained-patient\" }" +
-			"}";
-		Observation o = myFhirCtx.newJsonParser().parseResource(Observation.class, json);
+	@Nested
+	public class WithContainedIndexing {
+		@BeforeEach
+		public void enableContains() {
+			// we don't support chained or contained yet, but turn it on to test we don't blow up.
+			myDaoConfig.getModelConfig().setIndexOnContainedResources(true);
+			myDaoConfig.getModelConfig().setIndexOnContainedResourcesRecursively(true);
+		}
 
-		myObservationDao.create(o, mySrd).getId().toUnqualifiedVersionless();
+		@AfterEach
+		public void restoreContains() {
+			ModelConfig defaultModelConfig = new ModelConfig();
+			myDaoConfig.getModelConfig().setIndexOnContainedResources(defaultModelConfig.isIndexOnContainedResources());
+			myDaoConfig.getModelConfig().setIndexOnContainedResourcesRecursively(defaultModelConfig.isIndexOnContainedResourcesRecursively());
+		}
+		/**
+		 * We were throwing when indexing contained.
+		 * https://github.com/hapifhir/hapi-fhir/issues/3371
+		 */
+		@Test
+		public void ignoreContainedResources_noError() {
+			// given
+			String json =
+				"{" +
+					"\"resourceType\": \"Observation\"," +
+					"\"contained\": [{" +
+					"\"resourceType\": \"Patient\"," +
+					"\"id\": \"contained-patient\"," +
+					"\"name\": [{ \"family\": \"Smith\"}]" +
+					"}]," +
+					"\"subject\": { \"reference\": \"#contained-patient\" }" +
+					"}";
+			Observation o = myFhirCtx.newJsonParser().parseResource(Observation.class, json);
 
-		// no error.
+			myObservationDao.create(o, mySrd).getId().toUnqualifiedVersionless();
+
+			// no error.
+		}
 	}
+
 	
 
 	@Test
@@ -543,7 +559,6 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 		ArrayList<String> codes = toCodesContains(result.getExpansion().getContains());
 		assertThat(codes, containsInAnyOrder("childAAA", "childAAB"));
 	}
-
 
 	@Test
 	public void testExpandWithFilter() {

--- a/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java
+++ b/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchWithElasticSearchIT.java
@@ -125,6 +125,9 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 	public void beforePurgeDatabase() {
 		purgeDatabase(myDaoConfig, mySystemDao, myResourceReindexingSvc, mySearchCoordinatorSvc, mySearchParamRegistry, myBulkDataExportSvc);
 		myDaoConfig.setAdvancedLuceneIndexing(true);
+		// we don't support chained or contained yet, but turn it on to test we don't blow up.
+		myDaoConfig.getModelConfig().setIndexOnContainedResources(true);
+		myDaoConfig.getModelConfig().setIndexOnContainedResourcesRecursively(true);
 	}
 
 	@Override
@@ -499,6 +502,31 @@ public class FhirResourceDaoR4SearchWithElasticSearchIT extends BaseJpaTest {
 	private void assertObservationSearchMatches(String message, SearchParameterMap map, IIdType  ...iIdTypes) {
 		assertThat(message, toUnqualifiedVersionlessIdValues(myObservationDao.search(map)), containsInAnyOrder(toValues(iIdTypes)));
 	}
+
+	/**
+	 * We were throwing when indexing contained.
+	 * https://github.com/hapifhir/hapi-fhir/issues/3371
+	 */
+	@Test
+	public void ignoreContainedResources_noError() {
+	    // given
+		String json =
+			"{" +
+			"\"resourceType\": \"Observation\"," +
+			"\"contained\": [{" +
+			"\"resourceType\": \"Patient\"," +
+			"\"id\": \"contained-patient\"," +
+			"\"name\": [{ \"family\": \"Smith\"}]" +
+			"}]," +
+			"\"subject\": { \"reference\": \"#contained-patient\" }" +
+			"}";
+		Observation o = myFhirCtx.newJsonParser().parseResource(Observation.class, json);
+
+		myObservationDao.create(o, mySrd).getId().toUnqualifiedVersionless();
+
+		// no error.
+	}
+	
 
 	@Test
 	public void testExpandWithIsAInExternalValueSet() {

--- a/hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java
+++ b/hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java
@@ -48,7 +48,7 @@ public class ExtendedLuceneIndexData {
 
 	private <V> BiConsumer<String, V> ifNotContained(BiConsumer<String, V> theIndexWriter) {
 		return (s,v) -> {
-			// Ignore contained resources.
+			// Ignore contained resources for now.
 			if (!s.contains(".")) {
 				theIndexWriter.accept(s,v);
 			}

--- a/hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java
+++ b/hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedLuceneIndexData.java
@@ -28,6 +28,8 @@ import org.hibernate.search.engine.backend.document.DocumentElement;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.function.BiConsumer;
+
 /**
  * Collects our lucene extended indexing data.
  *
@@ -44,13 +46,22 @@ public class ExtendedLuceneIndexData {
 		this.myFhirContext = theFhirContext;
 	}
 
+	private <V> BiConsumer<String, V> ifNotContained(BiConsumer<String, V> theIndexWriter) {
+		return (s,v) -> {
+			// Ignore contained resources.
+			if (!s.contains(".")) {
+				theIndexWriter.accept(s,v);
+			}
+		};
+	}
 	public void writeIndexElements(DocumentElement theDocument) {
 		HibernateSearchIndexWriter indexWriter = HibernateSearchIndexWriter.forRoot(myFhirContext, theDocument);
 
-		// TODO MB Use RestSearchParameterTypeEnum to define templates.
-		mySearchParamStrings.forEach(indexWriter::writeStringIndex);
-		mySearchParamTokens.forEach(indexWriter::writeTokenIndex);
-		mySearchParamLinks.forEach(indexWriter::writeReferenceIndex);
+		ourLog.debug("Writing JPA index to Hibernate Search");
+
+		mySearchParamStrings.forEach(ifNotContained(indexWriter::writeStringIndex));
+		mySearchParamTokens.forEach(ifNotContained(indexWriter::writeTokenIndex));
+		mySearchParamLinks.forEach(ifNotContained(indexWriter::writeReferenceIndex));
 	}
 
 	public void addStringIndexData(String theSpName, String theText) {

